feat: Add How-To on managing application credentials

colder-is-better opened the pull request:
We explain what OpenStack application credentials are, and then we show how to manage them by providing specific examples. While at it, we add a new page regarding Keystone.

fghaas commented on docs/howto/openstack/keystone/app-creds.md:
````diff
@@ -0,0 +1,277 @@
++--------------+----------------------------------------------------------+
+```
+
+No matter how you choose to list or view application credentials, secrets **are never** displayed.
````
Emphasising the two words "are never" looks odd to me. Either drop the emphasis altogether, or if you must, emphasise just "never".

fghaas commented on docs/howto/openstack/keystone/app-creds.md:
````diff
@@ -0,0 +1,277 @@
+You may select specific columns...
+
+```console
+$ openstack application credential list -c ID -c Name -c Expires\ At
````
Somehow, I find `-c "Expires At"` (with quotes) more readable than the backslash escape.

fghaas commented on docs/howto/openstack/keystone/app-creds.md:
```diff
@@ -0,0 +1,277 @@
+
+## Restricted vs unrestricted credentials
+
+By default, application credentials are created as being *restricted*; that is why in the `openstack application credential show` outout above, the `unrestricted` parameter is set to `False`.
```
Typo. And also, I think that instead of the semicolon, you can just put two sentences. And I'd inject one more comma:
```suggestion
By default, application credentials are created as being *restricted*.
That is why, in the `openstack application credential show` output above, the `unrestricted` parameter is set to `False`.
```

fghaas commented on docs/howto/openstack/keystone/app-creds.md:
```diff
@@ -0,0 +1,277 @@
+
+By default, application credentials are created as being *restricted*; that is why in the `openstack application credential show` outout above, the `unrestricted` parameter is set to `False`.
+You cannot use restricted application credentials for Heat or Magnum, or for managing other application credentials or [trusts](https://docs.openstack.org/keystone/latest/user/trusts.html).
+If your application **has** to be able to perform such actions, and you accept the risks involved, you can always create *unrestricted* application credentials like this:
```
Drop the word "always" here.

fghaas commented on docs/howto/openstack/keystone/app-creds.md:
```diff
@@ -0,0 +1,277 @@
+
+In the example output above, the `unrestricted` parameter is set to `True`.
+
+> The restricted-by-default policy acts as a safeguard, so compromised application credentials cannot be used for creating other sets of application credentials.
```
Move this sentence up, so that it's part of the explanation of what differentiates restricted and unrestricted credentials (before the `--unrestricted` command example, that is).

fghaas commented on docs/howto/openstack/keystone/app-creds.md:
````diff
@@ -0,0 +1,277 @@
+source acRC
+```
+
+You can now use the `openstack` client as per usual.
````
"As usual" works; "per usual" would work too, but "as per usual" sounds positively odd to me.
```suggestion
You can now use the `openstack` client as usual.
```

fghaas commented on docs/howto/openstack/keystone/app-creds.md:
```diff
@@ -0,0 +1,277 @@
+
+* `OS_AUTH_URL`, which, in our example, is set to `https://fra1.citycloud.com:5000`,
+* `OS_AUTH_TYPE`, which should be set to `v3applicationcredential`,
+* `OS_APPLICATION_CREDENTIAL_ID`, which should be set to the value of `id` (as displayed in the table above),
```
Maybe mention that alternatively, one can also use `OS_APPLICATION_CREDENTIAL_NAME`, in which case one doesn't need the generated `id` at all?

fghaas commented on docs/howto/openstack/keystone/app-creds.md:
````diff
@@ -0,0 +1,277 @@
+No applicationcredential with a name or ID of 'goofy_dijkstra' exists.
+```
+
+Expired application credentials stay around until you explicitly delete them.
````
```suggestion
Expired application credentials stay around until you explicitly delete them, but can no longer be used for authentication.
```

fghaas reviewed (commented):
Good start; I have a few suggestions.

fghaas commented:
One more thing: please add another heading, "Troubleshooting", explaining the source of the `Application credentials cannot request a scope` error and its remedy.

colder-is-better commented:
Very nice! I have taken into account all of your suggestions and modified/augmented the prose accordingly. I have also added a new troubleshooting section.

fghaas commented on docs/howto/openstack/keystone/app-creds.md:
````diff
@@ -0,0 +1,310 @@
+| user_id      | e50719d594f84f16ad13f88da540f762                                                  |
++--------------+-----------------------------------------------------------------------------------+
+```
+
````
I think we should mention here that the `--expiration` argument format is an [ISO 8601](https://en.wikipedia.org/wiki/ISO_8601) date string (which implies that it can be abbreviated to just a date, and that one can add a timezone specifier if necessary).

colder-is-better replied:
Yes, added a note about that!

fghaas reviewed (commented):
Looks great now. One more tiny addition, and then I think this is good to merge,

fghaas commented on docs/howto/openstack/keystone/app-creds.md:
````diff
@@ -0,0 +1,314 @@
++--------------+-----------------------------------------------------------------------------------+
+```
+
+> The `--expiration` parameter accepts [ISO 8601](https://en.wikipedia.org/wiki/ISO_8601) strings, used for specifying date and time-related data.
+> Those strings may lack time data and carry just a date.
+> Should you wish, you may also add a timezone specifier.
````
The first sentence suffices:
```suggestion
> The `--expiration` parameter accepts [ISO 8601](https://en.wikipedia.org/wiki/ISO_8601) strings, used for specifying date and time-related data.
```

fghaas reviewed (approved):
Excellent, thanks!
